lib/pure/unicode.nim

```diff
@@ -1352,6 +1369,136 @@ proc isCombining*(c: Rune): bool {.rtl, extern: "nuc$1", procvar.} =
     (c >= 0x20d0 and c <= 0x20ff) or
     (c >= 0xfe20 and c <= 0xfe2f))
 
+proc swapCase*(s: string): string {.noSideEffect, procvar,
+  rtl, extern: "nuc$1".} =
+  ## Swaps the case of unicode characters in `s`
+  ##
+  ## Returns a new string such that the cases of all unicode characters
+  ## are swapped if possible
+
+  var
+    i = 0
+    lastIndex = 0
+    rune: Rune
+
+  result = newString(len(s))
+
+  while i < len(s):
+    lastIndex = i
+
+    fastRuneAt(s, i, rune)
+
+    if rune.isUpper():
+      rune = rune.toLower()
+    elif rune.isLower():
+      rune = rune.toUpper()
+
+    rune.fastToUTF8Copy(result, lastIndex)
+
+proc translate*(s: string, replacements: proc(key: string): string): string {.
```

Comment:
It's worth noting that this is different to the python `translate` function (although arguably more appropriately named than python's one).  Python's `translate` function works on characters and is essentially `str.replace` on steroids in that it allows multiple replacements in one go.  The important distinction between `str.replace` and `str.translate` is that you can do bidirectional replacing.  For example:

```python
import string
s = "abc"
print s.replace('a', 'c').replace('c', 'a') # Produces "aba"
print.s.translate(string.maketrans("ac", "ca")) # Produces "cba"
```

One simple use case (taken from [here](http://www.dotnetperls.com/translate-python)): a python program that applies a rot13 translation:
```python
import string
# Create translation table.
trans = string.maketrans("abcdefghijklmnopqrstuvwxyz",
    "nopqrstuvwxyzabcdefghijklm")

# Apply rot13 translation.
print("gandalf".translate(trans))
print("gandalf".translate(trans).translate(trans))
```

Output
```
tnaqnys
gandalf
```

I'm not sure how often this is used (I've used the python one a few times, but I could certainly understand it's equivalent being in `strmisc.nim`), but for the sake of #4218 I though it was worth mentioning that this wouldn't fill that gap in the table.  I also note that `translate` is only in the unicode module (which probably makes sense for this functionality) whereas the pythonic `translate` is just as useful for normal strings.